sql-plugin/src/main/scala/org/apache/spark/sql/rapids/ExternalSource.scala

```diff
@@ -21,15 +21,46 @@ import scala.util.{Failure, Success, Try}
 import com.nvidia.spark.rapids._
 
 import org.apache.spark.broadcast.Broadcast
-import org.apache.spark.sql.avro.{AvroFileFormat, AvroOptions}
 import org.apache.spark.sql.connector.read.{PartitionReaderFactory, Scan}
 import org.apache.spark.sql.execution.FileSourceScanExec
 import org.apache.spark.sql.execution.datasources.FileFormat
 import org.apache.spark.sql.sources.Filter
-import org.apache.spark.sql.v2.avro.AvroScan
 import org.apache.spark.util.{SerializableConfiguration, Utils}
 
+/**
+ * The subclass of AvroProvider imports spark-avro classes. This file should not imports
+ * spark-avro because `class not found` exception may throw if spark-avro does not exist at
+ * runtime. Details see: https://github.com/NVIDIA/spark-rapids/issues/5648
+ */
+trait AvroProvider {
+  /** If the file format is supported as an external source */
+  def isSupportedFormat(format: FileFormat): Boolean
+
+  def isPerFileReadEnabledForFormat(format: FileFormat, conf: RapidsConf): Boolean
+
+  def tagSupportForGpuFileSourceScan(meta: SparkPlanMeta[FileSourceScanExec]): Unit
+
+  /**
+   * Get a read file format for the input format.
+   * Better to check if the format is supported first by calling 'isSupportedFormat'
+   */
+  def getReadFileFormat(format: FileFormat): FileFormat
+
+  /**
+   * Create a multi-file reader factory for the input format.
+   * Better to check if the format is supported first by calling 'isSupportedFormat'
+   */
+  def createMultiFileReaderFactory(
+      format: FileFormat,
+      broadcastedConf: Broadcast[SerializableConfiguration],
+      pushedFilters: Array[Filter],
+      fileScan: GpuFileSourceScanExec): PartitionReaderFactory
+
+  def getScans: Map[Class[_ <: Scan], ScanRule[_ <: Scan]]
+}
+
 object ExternalSource {
+  val providerClassName = "org.apache.spark.sql.rapids.AvroSourceProvider"
 
   lazy val hasSparkAvroJar = {
     val loader = Utils.getContextOrSparkClassLoader
```

Comment:
We should do the same thing here as we are doing in GpuHiveOverrides. i.e. use the ShimLoader

Reply:
Done